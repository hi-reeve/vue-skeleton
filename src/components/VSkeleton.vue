<script setup lang="ts">
import { computed } from 'vue';

interface Props {
	height?: string;
	width?: string;
	animation?: 'pulse' | 'wave';
	circle?: boolean;
	as?: string;
}

const props = withDefaults(defineProps<Props>(), {
	circle: false,
	animation: 'pulse',
	height: '15px',
	width: '150px',
	as: 'div',
});

const classes = computed(() => ({
	skeleton: true,
	'skeleton--circle': props.circle,
	'animate-pulse': props.animation === 'pulse',
	'after:animate-wave': props.animation === 'wave',
}));
</script>

<template>
	<component :is="as" :class="classes"></component>
</template>

<style scoped lang="scss">
.skeleton {
	--height: v-bind(props.height);
	--width: v-bind(props.width);

	@apply relative block h-[var(--height)] w-[var(--width)] bg-zinc-200 cursor-progress after:inset-0 after:block after:bg-gradient-to-r  after:from-transparent after:via-white/50 after:to-transparent after:absolute  overflow-hidden after:-translate-x-full;

	&--circle {
		@apply rounded-full;
	}
}
</style>
