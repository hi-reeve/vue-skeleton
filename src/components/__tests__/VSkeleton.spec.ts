import { mount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import VSkeleton from '../VSkeleton.vue';
interface Props {
	height?: string;
	width?: string;
	animation?: 'pulse' | 'wave';
	circle?: boolean;
	as?: string;
}
const wrapperFactory = (
	props: Props = {
		circle: false,
		animation: 'pulse',
		height: '15px',
		width: '150px',
		as: 'div',
	},
) => {
	return mount(VSkeleton, {
		props: props,
	});
};

describe('VSkeleton Component', () => {
	it('should not null component', () => {
		expect(VSkeleton).toBeTruthy();
	});

	it('should mount the component', () => {
		const wrapper = wrapperFactory();

		expect(wrapper.classes()).toContain('skeleton');
	});

	it('should have circle', () => {
		const wrapper = wrapperFactory({
			circle: true,
			width: '48px',
			height: '48px',
		});

		expect(wrapper.classes()).toContain('skeleton--circle');
	});

	it('should be a span skeleton', () => {
		const wrapper = wrapperFactory({
			as: 'span',
		});

		expect(wrapper).toMatchSnapshot();
	});
});
