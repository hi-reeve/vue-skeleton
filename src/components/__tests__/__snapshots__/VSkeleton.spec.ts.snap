// Vitest Snapshot v1

exports[`VSkeleton Component > should be a span skeleton 1`] = `
VueWrapper {
  "__app": {
    "_component": {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <span
        class="skeleton animate-pulse"
        data-v-24715563=""
      />
    </div>,
    "_context": {
      "app": [Circular],
      "components": {
        "transition": {
          "name": "transition",
        },
        "transition-group": {
          "name": "transition-group",
        },
      },
      "config": {
        "compilerOptions": {},
        "errorHandler": undefined,
        "globalProperties": {},
        "isNativeTag": [Function],
        "optionMergeStrategies": {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": {},
      "emitsCache": WeakMap {},
      "mixins": [
        {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": {},
      "reload": [Function],
    },
    "_instance": {
      "a": null,
      "accessCache": {},
      "appContext": {
        "app": [Circular],
        "components": {
          "transition": {
            "name": "transition",
          },
          "transition-group": {
            "name": "transition-group",
          },
        },
        "config": {
          "compilerOptions": {},
          "errorHandler": undefined,
          "globalProperties": {},
          "isNativeTag": [Function],
          "optionMergeStrategies": {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": {},
        "emitsCache": WeakMap {},
        "mixins": [
          {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": {},
      "da": null,
      "data": {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": [
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
        ],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": {},
      "propsDefaults": {},
      "propsOptions": [
        {},
        [],
      ],
      "provides": {},
      "proxy": {},
      "refs": {
        "VTU_COMPONENT": {
          "hasOwnProperty": [Function],
        },
      },
      "render": [Function],
      "renderCache": [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": [],
        "effects": [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": {},
      "slots": {},
      "sp": null,
      "subTree": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": {
          "a": null,
          "accessCache": {},
          "appContext": {
            "app": [Circular],
            "components": {
              "transition": {
                "name": "transition",
              },
              "transition-group": {
                "name": "transition-group",
              },
            },
            "config": {
              "compilerOptions": {},
              "errorHandler": undefined,
              "globalProperties": {},
              "isNativeTag": [Function],
              "optionMergeStrategies": {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": {},
            "emitsCache": WeakMap {},
            "mixins": [
              {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": {},
          "ctx": {
            "animation": "pulse",
            "as": "span",
            "circle": false,
            "height": "15px",
            "width": "150px",
          },
          "da": null,
          "data": {},
          "devtoolsRawSetupState": {
            "classes": ComputedRefImpl {
              "__v_isReadonly": true,
              "__v_isRef": true,
              "_cacheable": true,
              "_dirty": false,
              "_setter": [Function],
              "_value": {
                "after:animate-wave": false,
                "animate-pulse": true,
                "skeleton": true,
                "skeleton--circle": false,
              },
              "dep": Set {
                ReactiveEffect {
                  "active": true,
                  "allowRecurse": true,
                  "deps": [
                    Set {
                      [Circular],
                    },
                    [Circular],
                  ],
                  "fn": [Function],
                  "onTrack": undefined,
                  "onTrigger": undefined,
                  "parent": undefined,
                  "scheduler": [Function],
                },
              },
              "effect": ReactiveEffect {
                "active": true,
                "computed": [Circular],
                "deps": [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "parent": undefined,
                "scheduler": [Function],
              },
            },
            "props": {
              "animation": "pulse",
              "as": "span",
              "circle": false,
              "height": "15px",
              "width": "150px",
            },
          },
          "directives": null,
          "ec": null,
          "effect": ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
          "emit": [Function],
          "emitsOptions": null,
          "emitted": null,
          "exposeProxy": {
            "hasOwnProperty": [Function],
          },
          "exposed": {
            "hasOwnProperty": [Function],
          },
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": null,
          "next": null,
          "parent": [Circular],
          "props": {
            "animation": "pulse",
            "as": "span",
            "circle": false,
            "height": "15px",
            "width": "150px",
          },
          "propsDefaults": {},
          "propsOptions": [
            {
              "animation": {
                "0": false,
                "1": true,
                "default": "pulse",
                "required": false,
                "type": [Function],
              },
              "as": {
                "0": false,
                "1": true,
                "default": "div",
                "required": false,
                "type": [Function],
              },
              "circle": {
                "0": true,
                "1": true,
                "default": false,
                "required": false,
                "type": [Function],
              },
              "height": {
                "0": false,
                "1": true,
                "default": "15px",
                "required": false,
                "type": [Function],
              },
              "width": {
                "0": false,
                "1": true,
                "default": "150px",
                "required": false,
                "type": [Function],
              },
            },
            [
              "height",
              "width",
              "animation",
              "circle",
              "as",
            ],
          ],
          "provides": {},
          "proxy": {
            "animation": "pulse",
            "as": "span",
            "circle": false,
            "height": "15px",
            "width": "150px",
          },
          "refs": {},
          "render": [Function],
          "renderCache": [],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": [],
            "effects": [
              ReactiveEffect {
                "active": true,
                "computed": ComputedRefImpl {
                  "__v_isReadonly": true,
                  "__v_isRef": true,
                  "_cacheable": true,
                  "_dirty": false,
                  "_setter": [Function],
                  "_value": {
                    "after:animate-wave": false,
                    "animate-pulse": true,
                    "skeleton": true,
                    "skeleton--circle": false,
                  },
                  "dep": Set {
                    ReactiveEffect {
                      "active": true,
                      "allowRecurse": true,
                      "deps": [
                        Set {
                          [Circular],
                        },
                        [Circular],
                      ],
                      "fn": [Function],
                      "onTrack": undefined,
                      "onTrigger": undefined,
                      "parent": undefined,
                      "scheduler": [Function],
                    },
                  },
                  "effect": [Circular],
                },
                "deps": [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "parent": undefined,
                "scheduler": [Function],
              },
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "parent": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": {
            "attrs": {},
            "emit": [Function],
            "expose": [Function],
            "slots": {},
          },
          "setupState": {
            "classes": {
              "after:animate-wave": false,
              "animate-pulse": true,
              "skeleton": true,
              "skeleton--circle": false,
            },
            "props": {
              "animation": "pulse",
              "as": "span",
              "circle": false,
              "height": "15px",
              "width": "150px",
            },
          },
          "slots": {},
          "sp": null,
          "subTree": {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": null,
            "component": null,
            "dirs": null,
            "dynamicChildren": [],
            "dynamicProps": [
              "class",
            ],
            "el": <span
              class="skeleton animate-pulse"
              data-v-24715563=""
            />,
            "key": null,
            "patchFlag": 8,
            "props": {
              "class": "skeleton animate-pulse",
            },
            "ref": null,
            "scopeId": "data-v-24715563",
            "shapeFlag": 1,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "span",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": {
            "__file": "/home/reeve/projects/vue-skeleton/src/components/VSkeleton.vue",
            "__hmrId": "24715563",
            "__scopeId": "data-v-24715563",
            "components": {},
            "name": "VSkeleton",
            "props": {
              "animation": {
                "0": false,
                "1": true,
                "default": "pulse",
                "required": false,
                "type": [Function],
              },
              "as": {
                "0": false,
                "1": true,
                "default": "div",
                "required": false,
                "type": [Function],
              },
              "circle": {
                "0": true,
                "1": true,
                "default": false,
                "required": false,
                "type": [Function],
              },
              "height": {
                "0": false,
                "1": true,
                "default": "15px",
                "required": false,
                "type": [Function],
              },
              "width": {
                "0": false,
                "1": true,
                "default": "150px",
                "required": false,
                "type": [Function],
              },
            },
            "render": [Function],
            "setup": [Function],
          },
          "u": null,
          "uid": 5,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <span
          class="skeleton animate-pulse"
          data-v-24715563=""
        />,
        "key": null,
        "patchFlag": 0,
        "props": {
          "as": "span",
          "ref": "VTU_COMPONENT",
        },
        "ref": {
          "f": false,
          "i": [Circular],
          "k": undefined,
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "__file": "/home/reeve/projects/vue-skeleton/src/components/VSkeleton.vue",
          "__hmrId": "24715563",
          "__scopeId": "data-v-24715563",
          "components": {},
          "name": "VSkeleton",
          "props": {
            "animation": {
              "0": false,
              "1": true,
              "default": "pulse",
              "required": false,
              "type": [Function],
            },
            "as": {
              "0": false,
              "1": true,
              "default": "div",
              "required": false,
              "type": [Function],
            },
            "circle": {
              "0": true,
              "1": true,
              "default": false,
              "required": false,
              "type": [Function],
            },
            "height": {
              "0": false,
              "1": true,
              "default": "15px",
              "required": false,
              "type": [Function],
            },
            "width": {
              "0": false,
              "1": true,
              "default": "150px",
              "required": false,
              "type": [Function],
            },
          },
          "render": [Function],
          "setup": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 4,
      "um": null,
      "update": [Function],
      "vnode": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": {
          "app": [Circular],
          "components": {
            "transition": {
              "name": "transition",
            },
            "transition-group": {
              "name": "transition-group",
            },
          },
          "config": {
            "compilerOptions": {},
            "errorHandler": undefined,
            "globalProperties": {},
            "isNativeTag": [Function],
            "optionMergeStrategies": {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": {},
          "emitsCache": WeakMap {},
          "mixins": [
            {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <span
          class="skeleton animate-pulse"
          data-v-24715563=""
        />,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 2,
    "component": [Function],
    "config": {
      "compilerOptions": {},
      "errorHandler": undefined,
      "globalProperties": {},
      "isNativeTag": [Function],
      "optionMergeStrategies": {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.36",
  },
  "__setProps": [Function],
  "componentVM": {
    "animation": "pulse",
    "as": "span",
    "circle": false,
    "height": "15px",
    "width": "150px",
  },
  "isDisabled": [Function],
  "rootVM": {},
  "wrapperElement": <span
    class="skeleton animate-pulse"
    data-v-24715563=""
  />,
}
`;
